README.md
# GW AWS Audit Tool
> NOTE: This is a specialized tool to help with actions to take during an audit of AWS usage.

[![Go Report Card](https://goreportcard.com/badge/GoodwayGroup/gw-aws-audit)](https://goreportcard.com/report/GoodwayGroup/gw-aws-audit)

## Basic Usage

Please see [the docs for details on the commands.](./docs/gw-aws-audit.md)

Useful for clearing **large S3 buckets (many millions of objects)**, identifying egress EBS volumes and tracking S3 spend.

```
$ gw-aws-audit help
NAME:
   gw-aws-audit - a collection of tools to audit AWS.

USAGE:
   gw-aws-audit [global options] command [command options] [arguments...]

AUTHOR:
   Derek Smith <dsmith@goodwaygroup.com>

COMMANDS:
   s3               S3 related commands
   rds              RDS related commands
   ec2              EC2 related commands
   sg               Security Group related commands
   iam              IAM related commands
   cw               CloudWatch related commands
   install-manpage  Generate and install man page
   version, v       Print version info
   help, h          Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --help, -h  show help (default: false)

COPYRIGHT:
   (c) 2020 Goodway Group
```

## Installation

### [`asdf` plugin](https://github.com/GoodwayGroup/asdf-gw-aws-audit)

Add plugin:

```
$ asdf plugin-add gw-aws-audit https://github.com/GoodwayGroup/asdf-gw-aws-audit.git
```

Install the latest version:

```
$ asdf install gw-aws-audit latest
```

### [Homebrew](https://brew.sh) (for macOS users)

```
brew tap GoodwayGroup/gw-aws-audit
brew install gw-aws-audit
```

### curl binary

```
$ curl https://i.jpillora.com/GoodwayGroup/gw-aws-audit! | bash
```

### [docker](https://www.docker.com/)
The compiled docker images are maintained on [GitHub Container Registry (ghcr.io)](https://github.com/orgs/GoodwayGroup/packages/container/package/gw-aws-audit).
We maintain the following tags:

- `edge`: Image that is build from the current `HEAD` of the main line branch.
- `latest`: Image that is built from the [latest released version](https://github.com/GoodwayGroup/gw-aws-audit/releases)
- `x.y.z` (versions): Images that are build from the tagged versions within Github.

```bash
docker pull ghcr.io/goodwaygroup/gw-aws-audit
docker run -v "$PWD":/workdir ghcr.io/goodwaygroup/gw-aws-audit --version
```

### man page

To install `man` page:

```
$ gw-aws-audit install-manpage
```

## Audit helper script

There is a bash helper script in the repo at [audit.sh](audit.sh). This tool is useful in running an audit across many regions at once.

```
$ ./audit.sh

audit.sh helper script for gw-aws-audit

Usage:
    audit.sh [gw-aws-audit commands]

Examples:
> This will run the 'gw-aws-audit sg detached' command for every region in the US (default)

    $ audit.sh sg detached

> This will run the 'gw-aws-audit ec2 stopped-hosts' for ONLY the us-west-2 region

    $ AWS_REGION=us-west-2 audit.sh ec2 stopped-hosts

> This will run the 'gw-aws-audit ec2 stopped-hosts' for every region in the EU

    $ REGION=eu audit.sh ec2 stopped-hosts

> This will run the 'gw-aws-audit cw monitoring' using a specific version of the tool.

    $ BIN_PATH=./bin/gw-aws-audit audit.sh cw monitoring

Note: REGION env values (default: US):
US: us-east-1 us-east-2 us-west-1 us-west-2
EU: eu-central-1 eu-west-1 eu-west-2 eu-west-3 eu-south-1 eu-north-1
AP: ap-east-1 ap-south-1 ap-northeast-3 ap-northeast-2 ap-southeast-1 ap-southeast-2 ap-northeast-1
CH: cn-north-1 cn-northwest-1
ROW: af-south-1 me-south-1 sa-east-2
ALL: All of the above combined

You can also set AWS_REGION and that will supersede the value of REGION
✔ Have fun!
```

### Command Categories

There are commands for `s3`, `ec2`, `rds`, `sg` and `cw`

**s3**
```
$ gw-aws-audit s3
...
COMMANDS:
   add-cost-tag                Add s3-cost-name to all S3 buckets
   metrics                     Get usage metrics
   clear-bucket, exterminatus  Clear all Objects within a given Bucket
```

**ec2**
```
$ gw-aws-audit ec2
...
COMMANDS:
   enhanced-monitoring  Produce report of Enhanced Monitoring enabled instances
   detached-volumes     List detached EBS volumes and snapshot counts
   stopped-hosts        List stopped EC2 hosts and associated EBS volumes
```

**rds**
```
$ gw-aws-audit rds
...
COMMANDS:
   enhanced-monitoring  Produce report of Enhanced Monitoring enabled instances
```

**cw**
```
$ gw-aws-audit cw
...
COMMANDS:
   enhanced-monitoring  Produce report of Enhanced Monitoring enabled EC2 & RDS instances
```

### Example Outputs

#### ec2 stopped-hosts
```
$ gw-aws-audit ec2 stopped-hosts
                INSTANCE ID          NAME            VOLUME                 SIZE (GB)  SNAPSHOTS  MIN SIZE (GB)  COSTS
                i-09e42474f22039e23  dummy-box-test
                                                     vol-0d4b4a7bc95a4b8e4          8          0              0  $0.80
                                                     vol-0cc0f6cd3c99bc1cc          8          0              0  $0.80
                TOTALS               1 INSTANCES     2 VOLUMES                  16 GB          0           0 GB  $1.60
```

#### ec2 detached-volumes
```
$ gw-aws-audit ec2 detached-volumes
           VOLUME                 SIZE (GB)  SNAPSHOTS  MIN SIZE (GB)  COSTS
           vol-0cc0f6cd3c99bc1cc          8          0              0  $0.80
   TOTALS  1 VOLUMES                   8 GB          0           0 GB  $0.80
```

#### cw enhanced-monitoring
```
$ gw-aws-audit cw enhanced-monitoring
Enhanced Metrics can add a cost. See: https://aws.amazon.com/cloudwatch/pricing/
Checking for EC2 Enhanced Monitoring

 NAME                                              INSTANCE ID
 master-us-east-1c.masters.us-east-1.gwdocker.com  i-041h4jk12jk23sd
 jumpbox12                                         i-02412sdfgsgdfgs
 analytics-prod                                    i-0a87d921n1rtasd
 EC2 INSTANCES                                     3


Checking for RDS Enhanced Monitoring

 DB INSTANCE                                ENGINE
 airflow-womp-ba-prod-v2                    postgres
 dashboard-reporting-production-instance-1  aurora-postgresql
 service-loloyol-production                 aurora-mysql
 DB INSTANCES                               3
```

#### s3 clear-bucket
```
$ gw-aws-audit s3 clear-bucket --bucket yolo
-- WARNING -- PAY ATTENTION -- FOR REALS --
This will delete ALL objects in yolo
-- THIS ACTION IS NOT REVERSIBLE --
Are you SUPER sure? [yolo]
Enter a value: yolo

Proceeding with batch delete for bucket: yolo
Pages: 198788 Listed: 198788000 Deleted: 198781000 Retries: 18373 DPS: 1921.64
```

#### s3 metrics
```
$ gw-aws-audit s3 metrics > out.csv
Starting metrics pull...
Bucket metric pull complete. Buckets: 207 Processed: 207
```

## Built With

* go v1.14+
* make
* [git-chglog](https://github.com/git-chglog/git-chglog)
* [goreleaser](https://goreleaser.com/install/)

## Deployment

Run `./release.sh $VERSION`

This will update docs, changelog, add the tag, push main and the tag to the repo. The `goreleaser` action will publish the binaries to the Github Release.

If you want to simulate the `goreleaser` process, run the following command:

```
$ curl -sL https://git.io/goreleaser | bash -s -- --rm-dist --skip-publish --snapshot
```

## Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

1. Fork the [GoodwayGroup/gw-aws-audit](https://github.com/GoodwayGroup/gw-aws-audit) repo
1. Use `go >= 1.16`
1. Branch & Code
1. Run linters :broom: `golangci-lint run`
    - The project uses [golangci-lint](https://golangci-lint.run/usage/install/#local-installation)
1. Commit with a Conventional Commit
1. Open a PR

## Versioning

We employ [git-chglog](https://github.com/git-chglog/git-chglog) to manage the [CHANGELOG.md](CHANGELOG.md). For the versions available, see the [tags on this repository](https://github.com/GoodwayGroup/gw-aws-audit/tags).

## Authors

* **Derek Smith** - [@clok](https://github.com/clok)

See also the list of [contributors](https://github.com/GoodwayGroup/gwvault/contributors) who participated in this project.

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

## Sponsors

[![goodwaygroup][goodwaygroup]](https://goodwaygroup.com)

[goodwaygroup]: https://s3.amazonaws.com/gw-crs-assets/goodwaygroup/logos/ggLogo_sm.png "Goodway Group"


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="v1.21.2"></a>
## [v1.21.2] - 2021-06-10
### Bug Fixes
- **deps:** update github.com/hako/durafmt commit hash to 5c1018a ([#60](https://github.com/GoodwayGroup/gw-aws-audit/issues/60))
- **deps:** update all non-major dependencies ([#59](https://github.com/GoodwayGroup/gw-aws-audit/issues/59))
- **deps:** update all non-major dependencies ([#58](https://github.com/GoodwayGroup/gw-aws-audit/issues/58))

### Ci
- update goreleaser token ([#62](https://github.com/GoodwayGroup/gw-aws-audit/issues/62))


<a name="v1.21.1"></a>
## [v1.21.1] - 2021-05-13
### Bug Fixes
- **deps:** update all non-major dependencies ([#57](https://github.com/GoodwayGroup/gw-aws-audit/issues/57))
- **deps:** update all non-major dependencies ([#55](https://github.com/GoodwayGroup/gw-aws-audit/issues/55))
- **deps:** update module github.com/aws/aws-sdk-go to v1.38.20 ([#54](https://github.com/GoodwayGroup/gw-aws-audit/issues/54))
- **deps:** update module github.com/aws/aws-sdk-go to v1.38.19 ([#52](https://github.com/GoodwayGroup/gw-aws-audit/issues/52))
- **deps:** update all non-major dependencies ([#51](https://github.com/GoodwayGroup/gw-aws-audit/issues/51))

### Chore
- **deps:** update alpine docker tag to v3.13.5 ([#53](https://github.com/GoodwayGroup/gw-aws-audit/issues/53))
- **release:** v1.21.1


<a name="v1.21.0"></a>
## [v1.21.0] - 2021-04-07
### Chore
- **docs:** updating docs for version v1.21.0
- **release:** v1.21.0

### Features
- **iam:** add delete access key action and refactor


<a name="v1.20.0"></a>
## [v1.20.0] - 2021-04-06
### Chore
- **release:** v1.20.0

### Features
- **iam:** add ability to disable console access


<a name="v1.19.6"></a>
## [v1.19.6] - 2021-04-05
### Bug Fixes
- **ci:** fix typo in makefile
- **deps:** update all non-major dependencies ([#50](https://github.com/GoodwayGroup/gw-aws-audit/issues/50))
- **deps:** update module github.com/aws/aws-sdk-go to v1.38.12 ([#49](https://github.com/GoodwayGroup/gw-aws-audit/issues/49))

### Chore
- **release:** v1.19.6


<a name="v1.19.5"></a>
## [v1.19.5] - 2021-04-01
### Bug Fixes
- **ci:** address docker release bug

### Chore
- **release:** v1.19.5


<a name="v1.19.4"></a>
## [v1.19.4] - 2021-04-01
### Chore
- udpate readme
- **make:** update makefile and release process
- **release:** v1.19.4

### Ci
- **docker:** add docker push to ghcr.io


<a name="v1.19.3"></a>
## [v1.19.3] - 2021-03-31
### Chore
- **docs:** updating docs for version v1.19.3

### Features
- **iam:** support viewing inline policies
- **release:** v1.19.3


<a name="v1.19.2"></a>
## [v1.19.2] - 2021-03-31
### Bug Fixes
- **iam:** user modify failed to pull all data

### Features
- **release:** v1.19.2


<a name="v1.19.1"></a>
## [v1.19.1] - 2021-03-31
### Bug Fixes
- **iam:** correct month calculation for treshold checks

### Features
- **release:** v1.19.1


<a name="v1.19.0"></a>
## [v1.19.0] - 2021-03-31
### Chore
- **docs:** updating docs for version v1.19.0

### Features
- **iam:** added commands for recently used and unused keys
- **release:** v1.19.0


<a name="v1.18.0"></a>
## [v1.18.0] - 2021-03-30
### Chore
- regenerate CHANGELOG.md
- update changelog config
- **docs:** updating docs for version v1.18.0

### Code Refactoring
- **aws client:** centralize client session management to reduce the number of instantiations of an AWS Session
- **iam:** cleanup render commands

### Features
- **iam:** add key deactivation bulk action
- **release:** v1.18.0


<a name="v1.17.0"></a>
## [v1.17.0] - 2021-03-26
### Bug Fixes
- **deps:** update module github.com/aws/aws-sdk-go to v1.38.6 ([#47](https://github.com/GoodwayGroup/gw-aws-audit/issues/47))

### Chore
- update README
- **docs:** updating docs for version v1.17.0

### Features
- **iam:** adding interactive management of users to the tool ([#48](https://github.com/GoodwayGroup/gw-aws-audit/issues/48))
- **release:** v1.17.0

### BREAKING CHANGE

The command `gw-aws-audit iam user-report` has been renamed to `gw-aws-audit iam report`


<a name="v1.16.0"></a>
## [v1.16.0] - 2021-03-22
### Chore
- **docs:** updating docs for version v1.16.0

### Features
- **iam:** user group and policy management
- **release:** v1.16.0


<a name="v1.15.2"></a>
## [v1.15.2] - 2021-03-17
### Bug Fixes
- **deps:** update github.com/hako/durafmt commit hash to 3a2c319 ([#45](https://github.com/GoodwayGroup/gw-aws-audit/issues/45))
- **deps:** update module github.com/aws/aws-sdk-go to v1.37.33 ([#46](https://github.com/GoodwayGroup/gw-aws-audit/issues/46))
- **goreleaser:** addressed type in binary naming

### Features
- **release:** v1.15.2


<a name="v1.15.1"></a>
## [v1.15.1] - 2021-03-17
### Chore
- **release:** update homebrew push

### Features
- **release:** v1.15.1


<a name="v1.15.0"></a>
## [v1.15.0] - 2021-03-15
### Chore
- **ci:** remove deprecated release commands
- **ci:** port to using golangci-lint and goreleaser github action
- **deps:** update kemba, awessession, cdocs, go-funk and aws-sdf-go
- **go.mod:** bump to go 1.16

### Features
- **release:** v1.15.0


<a name="v1.14.3"></a>
## [v1.14.3] - 2021-03-03
### Bug Fixes
- **deps:** update all non-major dependencies ([#40](https://github.com/GoodwayGroup/gw-aws-audit/issues/40))

### Chore
- **deps:** update module stretchr/testify to v1.7.0 ([#32](https://github.com/GoodwayGroup/gw-aws-audit/issues/32))
- **deps:** update module aws/aws-sdk-go to v1.36.28 ([#31](https://github.com/GoodwayGroup/gw-aws-audit/issues/31))
- **deps:** update module aws/aws-sdk-go to v1.36.18 ([#30](https://github.com/GoodwayGroup/gw-aws-audit/issues/30))
- **github actions:** add go proxy warming
- **renovate:** add extension for group:allNonMajor
- **renovate:** add gomodTidy option

### Features
- **release:** v1.14.3


<a name="v1.14.2"></a>
## [v1.14.2] - 2020-12-29
### Chore
- **deps:** update module aws/aws-sdk-go to v1.36.17
- **deps:** update module urfave/cli/v2 to v2.3.0

main.go
package main

import (
	"fmt"
	"log"
	"os"
	"runtime"
	"time"

	"github.com/GoodwayGroup/gw-aws-audit/ec2"
	"github.com/GoodwayGroup/gw-aws-audit/iam"
	"github.com/GoodwayGroup/gw-aws-audit/info"
	"github.com/GoodwayGroup/gw-aws-audit/rds"
	"github.com/GoodwayGroup/gw-aws-audit/s3"
	"github.com/GoodwayGroup/gw-aws-audit/sg"
	"github.com/clok/cdocs"
	"github.com/urfave/cli/v2"
)

var version string

func init() {
	cli.VersionPrinter = func(c *cli.Context) {
		_, _ = fmt.Fprintf(c.App.Writer, "%s %s (%s/%s)\n", info.AppName, version, runtime.GOOS, runtime.GOARCH)
	}
}

func main() {
	im, err := cdocs.InstallManpageCommand(&cdocs.InstallManpageCommandInput{
		AppName: info.AppName,
	})
	if err != nil {
		log.Fatal(err)
	}

	app := &cli.App{
		Name:     info.AppName,
		Version:  version,
		Compiled: time.Now(),
		Authors: []*cli.Author{
			{
				Name:  "Derek Smith",
				Email: "dsmith@goodwaygroup.com",
			},
		},
		Copyright:            "(c) 2020 Goodway Group",
		HelpName:             info.AppName,
		Usage:                "a collection of tools to audit AWS.",
		EnableBashCompletion: true,
		Commands: []*cli.Command{
			{
				Name:  "s3",
				Usage: "S3 related commands",
				Subcommands: []*cli.Command{
					{
						Name:  "add-cost-tag",
						Usage: "Add s3-cost-name to all S3 buckets",
						UsageText: `
Idempotent action that will add the ` + "`s3-cost-name`" + ` tag to ALL S3 buckets for a
given account.

The value will be the Bucket name.
`,
						Action: func(c *cli.Context) error {
							err := s3.AddCostTag()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
					{
						Name:  "metrics",
						Usage: "Get usage metrics",
						UsageText: `
Prints out a CSV report to STDOUT to help track usage across all buckets for a
given account.

Metrics per Bucket:

Objects (count)
Size (Bytes)
Size (GB)
Size (TB)
Bytes per Object
MB per Object
Has Cost Tag
`,
						Action: func(c *cli.Context) error {
							err := s3.GetBucketMetrics()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
					{
						Name:    "clear-bucket",
						Aliases: []string{"exterminatus"},
						Usage:   "Clear all Objects within a given Bucket",
						UsageText: `
Efficiently delete all objects within a bucket.

This process will run multiple paged deletes in parallel. It will handle API
throttling from AWS with an exponential backoff with retry. 
`,
						Flags: []cli.Flag{
							&cli.StringFlag{
								Name:     "bucket",
								Aliases:  []string{"b"},
								Usage:    "Bucket to clear",
								Required: true,
							},
						},
						Action: func(c *cli.Context) error {
							err := s3.ClearBucketObjects(c)
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
				},
			},
			{
				Name:  "rds",
				Usage: "RDS related commands",
				Subcommands: []*cli.Command{
					{
						Name:  "enhanced-monitoring",
						Usage: "Produce report of Enhanced Monitoring enabled instances",
						Action: func(c *cli.Context) error {
							err := rds.ListMonitoringEnabled()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
					{
						Name:  "public",
						Usage: "Produce report of instances that have public interfaces attached",
						UsageText: `
Produces a report that displays a list RDS servers that are configured as Publicly Accessible.

The report contains:

DB INSTANCE:
    - Name of the instance

ENGINE:
    - RDS DB engine

SECURITY GROUPS:
    - Security Group ID
    - Security Group Name
    - Inbound Port
    - CIDR rules applied to the Port
`,
						Action: func(c *cli.Context) error {
							err := rds.ListPublicInterfaces()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
				},
			},
			{
				Name:  "ec2",
				Usage: "EC2 related commands",
				Subcommands: []*cli.Command{
					{
						Name:  "enhanced-monitoring",
						Usage: "Produce report of Enhanced Monitoring enabled instances",
						Action: func(c *cli.Context) error {
							err := ec2.ListMonitoringEnabled()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
					{
						Name:  "detached-volumes",
						Usage: "List detached EBS volumes and snapshot counts",
						Action: func(c *cli.Context) error {
							err := ec2.ListDetachedVolumes()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
					{
						Name:  "stopped-hosts",
						Usage: "List stopped EC2 hosts and associated EBS volumes",
						Action: func(c *cli.Context) error {
							err := ec2.ListStoppedHosts()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
					{
						Name:  "pem-keys",
						Usage: "List instances and PEM key used at time of creation",
						Action: func(c *cli.Context) error {
							err := ec2.ListPemKeyUsage()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
				},
			},
			{
				Name:  "sg",
				Usage: "Security Group related commands",
				Subcommands: []*cli.Command{
					{
						Name:  "detached",
						Usage: "generate a report of all Security Groups that are NOT attached to an instance",
						UsageText: `
This command will scan the EC2 NetworkInterfaces to determine what
Security Groups are NOT attached/assigned in AWS.
`,
						Action: func(c *cli.Context) error {
							err := sg.ListDetachedSecurityGroups()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
					{
						Name:  "attached",
						Usage: "generate a report of all Security Groups that are attached to an instance",
						UsageText: `
This command will scan the EC2 NetworkInterfaces to determine what
Security Groups are attached/assigned in AWS.
`,
						Action: func(c *cli.Context) error {
							err := sg.ListAttachedSecurityGroups()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
					{
						Name:  "cidr",
						Usage: "generate a report comparing SG rules with input CIDR blocks",
						UsageText: `
$ gw-aws-audit sg cidr --allowed 10.176.0.0/16,10.175.0.0/16 --alert 174.0.0.0/8,1.2.3.4/32

This command will generate a report detecting the port to CIDR mapping rules 
for attached Security Groups. 

A list of Approved CIDRs is required. This is typically the CIDR block associated
with your VPC.
`,
						Flags: []cli.Flag{
							&cli.StringFlag{
								Name:     "approved",
								Aliases:  []string{"a"},
								Usage:    "CIDR blocks that are approved (csv)",
								Required: true,
							},
							&cli.StringFlag{
								Name:    "warn",
								Aliases: []string{"w"},
								Usage:   "CIDR blocks that will cause a warning (csv)",
								Value:   "204.0.0.0/8",
							},
							&cli.StringFlag{
								Name:    "alert",
								Aliases: []string{"b"},
								Usage:   "CIDR blocks that will cause an alert (csv)",
								Value:   "174.0.0.0/8",
							},
							&cli.StringFlag{
								Name:    "ignore-ports",
								Aliases: []string{"p"},
								Usage:   "Ports that can be ignored (csv)",
								Value:   "80,443,3,4,3-4",
							},
							&cli.StringFlag{
								Name:  "ignore-protocols",
								Usage: "Protocols to ignore. Can be tcp,udp,icmp (csv)",
							},
							&cli.BoolFlag{
								Name:  "all",
								Usage: "Process ALL Security Groups, not just attached",
								Value: false,
							},
						},
						Action: func(c *cli.Context) error {
							err := sg.GenerateCIDRReport(c)
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
					{
						Name:  "port",
						Usage: "generate a report comparing SG rules with input CIDR blocks on a specific port",
						UsageText: `
$ gw-aws-audit sg ports --ports 22,3306 --allowed 10.176.0.0/16,10.175.0.0/16 --alert 174.0.0.0/8,1.2.3.4/32

This command will generate a report for a set of PORTS for attached Security Groups.

A list of Approved CIDRs is required. This is typically the CIDR block associated
with your VPC.
`,
						Flags: []cli.Flag{
							&cli.StringFlag{
								Name:    "ports",
								Aliases: []string{"p"},
								Usage:   "Ports to generate report on (csv)",
								Value:   "22",
							},
							&cli.StringFlag{
								Name:     "approved",
								Aliases:  []string{"a"},
								Usage:    "CIDR blocks that are approved (csv)",
								Required: true,
							},
							&cli.StringFlag{
								Name:    "warn",
								Aliases: []string{"w"},
								Usage:   "CIDR blocks that will cause a warning (csv)",
								Value:   "204.0.0.0/8",
							},
							&cli.StringFlag{
								Name:    "alert",
								Aliases: []string{"b"},
								Usage:   "CIDR blocks that will cause an alert (csv)",
								Value:   "174.0.0.0/8",
							},
							&cli.StringFlag{
								Name:  "ignore-protocols",
								Usage: "Protocols to ignore. Can be tcp,udp,icmp (csv)",
							},
							&cli.BoolFlag{
								Name:  "all",
								Usage: "Process ALL Security Groups, not just attached",
								Value: false,
							},
						},
						Action: func(c *cli.Context) error {
							err := sg.GeneratePortReport(c)
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
					{
						Name:  "amazon",
						Usage: "generate a report of allow SG with rules mapped to known AWS IPs",
						UsageText: `
This method loads the current version of https://ip-ranges.amazonaws.com/ip-ranges.json
and compares the CIDR blocks against all Security Groups.
`,
						Action: func(c *cli.Context) error {
							err := sg.GenerateExternalAWSIPReport()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
					{
						Name:    "direct-ip-mapping",
						Aliases: []string{"dim"},
						Usage:   "generate report of Security Groups with direct mappings to EC2 instances",
						UsageText: `
This method will generate a report comparing all Security Groups with all 
EC2 instances to determine where you have a direct IP mapping.

This will note Internal and External IP usage as well.
`,
						Action: func(c *cli.Context) error {
							err := sg.GenerateMappedEC2Report()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
				},
			},
			{
				Name:  "iam",
				Usage: "IAM related commands",
				Subcommands: []*cli.Command{
					{
						Name:  "user",
						Usage: "Set of commands to take action on AWS Users",
						Subcommands: []*cli.Command{
							iam.ActionUserReport,
							iam.ActionUserModify,
							iam.ActionUserPermissions,
							iam.ActionUserKeys,
						},
					},
					{
						Name:      "keys",
						Usage:     "Set of commands to take action on AWS Access Keys",
						UsageText: "",
						Subcommands: []*cli.Command{
							iam.ActionKeysDeactivate,
							iam.ActionKeysDelete,
							iam.ActionKeysRecent,
							iam.ActionKeysUnused,
						},
					},
					iam.ActionDeprecatedUserReport,
				},
			},
			{
				Name:  "cw",
				Usage: "CloudWatch related commands",
				Subcommands: []*cli.Command{
					{
						Name:  "enhanced-monitoring",
						Usage: "Produce report of Enhanced Monitoring enabled EC2 & RDS instances",
						Action: func(c *cli.Context) error {
							fmt.Println("Enhanced Metrics can add a cost. See: https://aws.amazon.com/cloudwatch/pricing/")
							fmt.Printf("Checking for EC2 Enhanced Monitoring\n\n")
							err := ec2.ListMonitoringEnabled()
							if err != nil {
								return cli.Exit(err, 2)
							}
							fmt.Printf("\n\nChecking for RDS Enhanced Monitoring\n\n")
							err = rds.ListMonitoringEnabled()
							if err != nil {
								return cli.Exit(err, 2)
							}
							return nil
						},
					},
				},
			},
			im,
			{
				Name:    "version",
				Aliases: []string{"v"},
				Usage:   "Print version info",
				Action: func(c *cli.Context) error {
					fmt.Printf("%s %s (%s/%s)\n", info.AppName, version, runtime.GOOS, runtime.GOARCH)
					return nil
				},
			},
		},
	}
	app.EnableBashCompletion = true

	if os.Getenv("DOCS_MD") != "" {
		docs, err := cdocs.ToMarkdown(app)
		if err != nil {
			panic(err)
		}
		fmt.Println(docs)
		return
	}

	if os.Getenv("DOCS_MAN") != "" {
		docs, err := cdocs.ToMan(app)
		if err != nil {
			panic(err)
		}
		fmt.Println(docs)
		return
	}

	err = app.Run(os.Args)
	if err != nil {
		log.Fatal(err)
	}
}
